proposals/context.md

```diff
@@ -0,0 +1,265 @@
+
+Frameworks and libraries often provide mechanisms for this, these can range from the simple Context implementation available in React, to more complex Dependency Injection frameworks. Web components need a similar protocol in order to solve this problem.
+
+# Goals
+
+- Allow elements in the DOM to retrieve data based on their contextual position in the DOM
+- Alleviate the problem of 'prop drilling'
+- Simple API that is easily implemented in any framework / library
+- Synchronous protocol, while supporting asynchronous patterns
+- Support single or multiple delivery of context values
+
+# Non-Goals
+
+**Context API !== Dependency Injection Framework**
+
+The Context API does not intend to cover all cases and forms of Dependency Injection. It does not specify constructor, factory or property injection patterns. Its only goal is to formalize the pattern of sharing data across the hierarchy in the DOM, specifically avoiding 'prop drilling' type scenarios. Dependency Injection patterns could be implemented using this protocol, but this is not the goal and should remain explicitly outside the scope of Context API for simplicity.
+
+**Context API is not a state management alternative**
+
+State management libraries often need to perform similar behaviors to the problems that Context API helps to solve. An element deep in the DOM tree made need access to some state, and may need to respond to that state being changed. While state management could be built using the Context API, it is not a primary goal of Context API to solve this problem.
+
+# Details
+
+## The `request-context` event
+
+Components which wish to receive some data from their ancestors should initiate the request by firing a composed, bubbling, `request-context` Event, with a `callback` property.
+
+Typescript interface:
+
+```typescript
+interface RequestContextEvent extends Event {
+  /**
+   * The name of the context that is requested
+   */
+  readonly name: T;
+  /**
+   * A boolean indicating if the context should only be provided once.
+   */
+  readonly once: boolean;
+  /**
+   * A callback which a provider of this named callback should invoke.
+   */
+  readonly callback: RequestContextCallback<ContextTypeMap[T]>;
+}
+```
+
+A full typescript definition for this event and its associated types can be found in the [Definitions](#definitions) section at the end of this document.
+
+## Context Providers
+
+A context provider will satisfy a `request-context` event, passing the `callback` the requested data whenever the data changes. A provider will attach an event listener to the DOM tree to catch the event, and if it will be able to satisfy the request will call `stopPropagation` on the event.
```

Comment:
Should we tighten the language and say that in order to be compliant with the protocol, a listener _must_ call `stopPropagation()` if it provides a value now, or will in the future? I can't think of where it would make sense to have multiple providers.

Reply:
Done in bccaeac